### DISCUSSION TITLE
Filebeat third-party module dashboard intergrations

### DISCUSSION THREAD

**Discussion by SuiNom** [Security-Onion-Solutions/securityonion#5778] [Unsupported Versions]
Hi Folks,

I came across [this discussion post](https://github.com/Security-Onion-Solutions/securityonion/discussions/5596), but had no luck with the posted answer. ([see my post for full details](https://github.com/Security-Onion-Solutions/securityonion/discussions/5596#discussioncomment-1401832))

I was instructed to try again (and start a new discussion) but double check the filebeat versions I was downloading matches that of the Elastic stack, which in hindsight is what I should have done in the first place, so...

1.  Check version of filebeat in use:
```
[root@security-onion-0 ~]# docker exec -it so-filebeat filebeat version
filebeat version 7.13.4 (amd64), libbeat 7.13.4 [1907c246c8b0d23ae4027699c44bf3fbef57f4a4 built 2021-07-14 18:42:41 +0000 UTC]
```
2. Download the relevant version of filbeat from [here](https://artifacts.elastic.co/downloads/beats/filebeat/filebeat-7.13.4-linux-x86_64.tar.gz)
3. extract with `➜  Downloads tar -xvf filebeat-7.13.4-linux-x86_64.tar.gz` - Dashboards seem to be under `filebeat-7.13.4-linux-x86_64/kibana/7/dashboard/`
4. Import the dashboard (I tried `Filebeat-O365-Audit.json` & `filebeat-aws-cloudtrail.json`) via `Kibana > Stack Management > Saved Objects > Import`
5. Both of these resulted in the same errors listed in my [first post](https://github.com/Security-Onion-Solutions/securityonion/discussions/5596#discussioncomment-1401832), but I'll put them here for the search engine linkages. Error:
```
Sorry, there was an error
Saved objects file format is invalid and cannot be imported.
```
![2021-09-29_11-57_1](https://user-images.githubusercontent.com/22012477/135255961-3020a826-0ce7-4b5b-86dc-e1102cd83551.png)

I'm totally expecting for this to be user error, but if possible I'd rather not re-invent the dashboards that already exist for the more common use cases like O365, AWS CloudTrail, etc.

All help appreciated!

Cheers,
Jamie!

**Comment by dougburks** [comment]
Have you tried googling `Saved objects file format is invalid and cannot be imported`?  Here are some of the results which might be relevant:
https://github.com/elastic/kibana/issues/50753
https://discuss.elastic.co/t/manually-loading-index-patterns-and-dashboards-into-kibana-after-exporting-from-beats/220942

**Reply by SuiNom** [reply]
Thanks @dougburks,

You're Google-Fu style is obviously stronger than my DuckDuckGo-Fu style. You're second link has gotten me a step closer with the following jq command: `➜  dashboard jq -c '.objects[]' Filebeat-O365-Audit.json > Filebeat-O365-Audit.ndjson`.
This gives me an apparently valid .ndjson file that I can actually import into Kibana:
![2021-10-07_15-52](https://user-images.githubusercontent.com/22012477/136411078-3f68590a-7ebd-4f8b-9aec-78531ea97fb4.png)
Ignoring the notice about pattern conflicts, I had prior to this, created a `*:so-0365-*` Index Pattern that could be used (while I was tinkering with building my own dashboards, but my dislike for re-inventing wheels is driving me in this direction).

This gets me successfully importing some assets (again, ignore the `8 overwritten` notice - I re-imported to grab screenshots because I was impatient the first time :roll_eyes:, more haste, less speed...)
![2021-10-07_15-53](https://user-images.githubusercontent.com/22012477/136411817-af1b4873-09d0-4cea-b0b7-139ed265261d.png)

So viewing the dashboards, it's clear _something_ is working, because the `Total audit events` widget is populated, however that is about as much as we have working so far.

Incoming screenshot flood:
![2021-10-07_16-08](https://user-images.githubusercontent.com/22012477/136412786-ff990960-b26e-4ea8-953b-bf822049ba8c.png)
![2021-10-07_16-14_1](https://user-images.githubusercontent.com/22012477/136414018-32e90f9a-79e3-45a5-bc8e-986e2684b854.png)
![2021-10-07_16-14](https://user-images.githubusercontent.com/22012477/136414023-c2f234ed-ea0e-483c-806d-652f19fdf1ce.png)
![2021-10-07_16-13_1](https://user-images.githubusercontent.com/22012477/136414024-0701cf7b-65a2-4036-88fe-7932af922a7c.png)
![2021-10-07_16-13](https://user-images.githubusercontent.com/22012477/136414026-c808fc4d-477b-4574-a42d-bfe19dc1cdf4.png)

This is as much as I have time for today, but I do intend to try and unpick these errors and resolve them if possible, thought the map error regarding a missing field I'm much less hopeful for.

Supporting Filebeat modules was huge, and I'm so glad it was added, if we can forge a bit of a path that maybe makes the dashboard mobility side a little simpler too, I think it'd be amazing!

Cheers,
Jamie!

**Reply by SuiNom** [reply]
Okay, the DLP widget seems to be working too
![2021-10-07_16-59](https://user-images.githubusercontent.com/22012477/136421284-159b1823-8fae-4aa6-b1c1-fe3f40545894.png)
!

**Reply by SuiNom** [reply]
Okay, spurred on by a coffee and a hint of success I've now gotten most of the dashboard working. Turns out each "Visualization" only needed `.keyword` appended to each split series bucket field. The where and why-for's of the `.keyword` field escape me, but it gives me much prettier results:
![2021-10-07_17-20](https://user-images.githubusercontent.com/22012477/136425287-25373e66-6e93-4bb6-b83f-445d82520f79.png)

![2021-10-07_17-08](https://user-images.githubusercontent.com/22012477/136426234-72081c27-cb9c-42fb-9f4e-84393244d4f2.png)
![2021-10-07_17-12](https://user-images.githubusercontent.com/22012477/136426253-7dcec532-b3f2-44ed-acc9-452587ebb7ae.png)


There are some things that I need to spend some more time on, such as why the `Top users by authentication failures` visualisation is showing successes, but I'm intuiting that's a small problem.

Kibana Maps are still a bit of a dark art to me, For some reason I can't get any results regardless of geo-spatial field selected, but I'll take most of this as a win so far...

Still a work in progress.

**Reply by SuiNom** [reply]
Okay, I've been down the rabbit hole for maps, and I've come back with a kludge that is _okay_ for now:
![image](https://user-images.githubusercontent.com/22012477/136894623-df1f9412-74c9-4152-85c9-3b605476b90c.png)
As none of the `geo_point` fields (`destination_geo.location`, `geoip.location`, `source_geo.location`) within the `*:so-o365-*` index pattern are actually populated, I've added a "Choropleth" layer to the map - which uses the World Countries boundary layer, and joins the two alpha ISO code with the `source.geo.country_iso_code.keyword` field, to represent at least countries on a map.
![image](https://user-images.githubusercontent.com/22012477/136898198-c2cda56b-03e5-49ca-9519-8aabb4e27fed.png)

**The underlying problem appears to be that the longitude and latitude fields (`source.geo.location.lat`, `source.geo.location.lon`) are `float` fields, and from what I can see they aren't combined into a `geo_point` type field at any point.**

I've tried to dig into this a little and found some relevant material:
https://discuss.elastic.co/t/visualization-of-lat-lon-coordinates-on-kibanas-graph/106177 (old but seems relevant)
https://www.elastic.co/guide/en/elasticsearch/reference/7.x/geo-point.html

There seems to be a general indication that an index template is the way to go to combine the lon/lat fields into a `geo_point` field, but at this point I'm out of my depth, I've poked around with the `so-common` template, but I'm not certain on the ramifications of changing anything, and given this seems to be the index template for anything `*:so-*`, I'm not too keen to poke it with a stick...

There also seems to be the option of a custom ingest pipeline, but this is ingesting using the o365 third party filebeat module, and I don't actually see those in `/opt/so/conf/elasticsearch/ingest/` as mentioned [here](https://docs.securityonion.net/en/2.3/elasticsearch.html#parsing) - and even if I did I'd be awhile figuring out how to alter it in the correct way to combine two `float`s into a `geo_point`.

Hopefully someone with a bit more ingest pipline/index template experience can weigh in?

Cheers,
Jamie.
